### ISSUE TITLE
[BUG]: `armory.art_experimental.attacks.carla_obj_det_adversarial_patch.CARLAAdversarialPatchPyTorch` does not zero patch gradients before backward

### ISSUE THREAD

**Issue by dxoigmn** [twosixlabs/armory#1950]
### Description of the bug

Line 117 below zero's the estimator's gradients (of which there should be none), but it does not zero the gradients of `self._patch`. This causes the attack to accumulate gradients between iterations, which is probably not the intended behavior.

https://github.com/twosixlabs/armory/blob/3fd07b4b5e57368f171d940e9601a2016fcc532e/armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py#L117-L119

### Steps To Reproduce

1. Add the following code just before `loss.backward`:
```python
if self._patch.grad is not None:
    print(self._patch.grad.min(), self._patch.grad.max())
```
2. Observe that the printed gradient values are not zero.

**Comment by swsuggs** [comment]
Hi Cory, thanks for bringing this up.  This attack was modeled after the parent [ART class](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/7691b390e99eeedbb0bf5ae8ea6a6e569edae995/art/attacks/evasion/adversarial_patch/adversarial_patch_pytorch.py#L189-L191) which also does not clear the patch gradients.  I agree it seems incorrect, and a small test shows that the attack converges faster if the gradients are cleared.

**Comment by swsuggs** [comment]
Fixed in #1969.  Note for posterity: it was observed that the attack actually converges faster when using PGD if the gradients are not cleared but are allowed to accumulate.  The gradients are now cleared if using the Adam optimizer for the attack.
